contracts/src/CrossChainApplications/NativeTokenBridge/ERC20TokenSource.sol

```diff
@@ -0,0 +1,174 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "./IERC20TokenSource.sol";
+import "../../Teleporter/ITeleporterMessenger.sol";
+import "../../Teleporter/ITeleporterReceiver.sol";
+import "../../Teleporter/SafeERC20TransferFrom.sol";
+
+contract ERC20TokenSource is
+    ITeleporterReceiver,
+    IERC20TokenSource,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20;
+
+    uint256 public constant MINT_NATIVE_TOKENS_REQUIRED_GAS = 150_000; // TODO this is a placeholder
+    bytes32 public immutable currentBlockchainID;
+    bytes32 public immutable destinationBlockchainID;
+    address public immutable nativeTokenDestinationAddress;
+    address public immutable erc20ContractAddress;
+
+    // Used for sending an receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 destinationBlockchainID_,
+        address nativeTokenDestinationAddress_,
+        address erc20ContractAddress_
+    ) {
+        currentBlockchainID = WarpMessenger(
+            0x0200000000000000000000000000000000000005
+        ).getBlockchainID();
```

Comment:
We might want to remove the call to the Warp precompile from the constructor for similar reasons to https://github.com/ava-labs/teleporter/pull/64.

Say we've successfully deployed the destination contract, providing it with the expected address of the source. But on the source chain, the Warp precompile hasn't been enabled, so the deployment tx reverts and increments the deployer account nonce. The destination contract will then have an address for the source that can never be deployed to.

Reply:
I don't think we need to worry about this here. The requirement for these contracts is that a pair of two contracts need to know the address of the other. There's no requirement for the addresses to be the same across all chains, so its not a big deal if a deployment reverts, you can just deploy at a different address.

Reply:
Good point. I suppose it'd be difficult to put the contracts in a bad enough state before its counterpart is deployed that we couldn't just start from scratch.